File: services/api/verge_api/routes/watch.py
"""Continuous WatchLoop control — start / stop / status (product heartbeat)."""

from __future__ import annotations

from fastapi import APIRouter, Request
from pydantic import BaseModel, Field

from ..watch_loop import controller

router = APIRouter(tags=["watch"])


class WatchStartBody(BaseModel):
    intervalS: float | None = Field(default=None, ge=1.0, le=60.0)
    vision: bool | None = None
    voice: bool | None = None
    sensors: bool | None = None
    fuse: bool | None = None
    cognee: bool | None = None


def _legs_from_body(body: WatchStartBody) -> dict[str, bool] | None:
    legs: dict[str, bool] = {}
    for key in ("vision", "voice", "sensors", "fuse", "cognee"):
        val = getattr(body, key)
        if val is not None:
            legs[key] = val
    return legs or None


@router.get("/watch/status")
def watch_status() -> dict:
    return controller.public_status()


@router.post("/watch/start")
def watch_start(body: WatchStartBody, request: Request) -> dict:
    controller.bind_app(request.app)
    status = controller.start(interval_s=body.intervalS, legs=_legs_from_body(body))
    return {"ok": True, "watch": status}


@router.post("/watch/stop")
def watch_stop(request: Request) -> dict:
    controller.bind_app(request.app)
    status = controller.stop()
    return {"ok": True, "watch": status}

File: apps/console/src/api/watch.ts
/** Continuous WatchLoop — start/stop/status. No frontend fiction. */

export interface WatchStatus {
  running: boolean;
  mode?: string;
  scenarioId?: string | null;
  scenarioLabel?: string | null;
  coach?: string | null;
  startedAt?: string | null;
  intervalS: number;
  ticks: number;
  lastTickAt?: string | null;
  lastError?: string | null;
  legs: Record<string, boolean>;
  counts: Record<string, number>;
  last?: Record<string, unknown>;
}

export async function fetchWatchStatus(): Promise<WatchStatus> {
  const res = await fetch('/api/watch/status');
  if (!res.ok) throw new Error(`watch status ${res.status}`);
  return res.json() as Promise<WatchStatus>;
}

export async function startWatch(opts?: {
  intervalS?: number;
  vision?: boolean;
  voice?: boolean;
  sensors?: boolean;
  fuse?: boolean;
  cognee?: boolean;
}): Promise<WatchStatus> {
  const res = await fetch('/api/watch/start', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(opts ?? {}),
  });
  if (!res.ok) throw new Error(`watch start ${res.status}`);
  const body = (await res.json()) as { watch: WatchStatus };
  return body.watch;
}

export async function stopWatch(): Promise<WatchStatus> {
  const res = await fetch('/api/watch/stop', { method: 'POST' });
  if (!res.ok) throw new Error(`watch stop ${res.status}`);
  const body = (await res.json()) as { watch: WatchStatus };
  return body.watch;
}
